### PR TITLE
`search_model` added to index settings

### PR DESCRIPTION
**Pull request by vicilliar**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
An index only has `model`, which is used to encode vectors for both add documents and search.

* **What is the new behavior (if this is a feature change)?**
Index settings defaults can now have `search_model` and `search_model_properties`, which will determine the model used for encoding search queries.
Notes (user-facing):
- If not explicitly set, search_model will always assimilate value of model.
- search_model cannot be set without setting model.
- search_model dimensions must match model dimensions.

Notes (internal):
- new module `tensor_search/create_index.py` created to contain all utils for creating indexes.
- Tests for index settings and IndexInfo all updated to include search_model.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
yes (but still adding more)

* **Related Python client changes** (link commit/PR here)
to follow

* **Related documentation changes** (link commit/PR here)
to follow

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/tensor_search/validation.py [hunk 1]
```diff
@@ -742,3 +742,25 @@ def validate_model_properties_no_model(model_properties: dict):
...
+    3. if search_model_properties is given, then search_model must be given as well.
+    """
+    model_name = index_settings[IndexSettingsField.index_defaults].get(IndexSettingsField.model)
+    model_properties = index_settings[IndexSettingsField.index_defaults].get(IndexSettingsField.model_properties)
+    search_model_name = index_settings[IndexSettingsField.index_defaults].get(IndexSettingsField.search_model)
+    search_model_properties = index_settings[IndexSettingsField.index_defaults].get(IndexSettingsField.search_model_properties)
+
+
+    if search_model_name is not None and model_name is None:
+        raise UnknownModelError(f"`search_model` cannot be specified without also specifying `model`. Received search_model={search_model_name} but found no `model`")
```
Comment (farshidz) [inline src/marqo/tensor_search/validation.py]: All three look like they should be `InvalidArgError` as they are user input errors, not inference errors?

File: src/marqo/tensor_search/backend.py [hunk 2]
```diff
@@ -47,12 +47,21 @@ def get_index_info(
...
+    # Identify index `model` from mapping metadata
     if "model" in res[index_name]["mappings"]["_meta"]:
         model_name = res[index_name]["mappings"]["_meta"]["model"]
     else:
         raise errors.NonTensorIndexError(
             "get_index_info: couldn't identify embedding model name "
             F"in index mappings! Mapping: {res}")
+    
+    # Identify index `search_model` from mapping metadata
+    if "search_model" in res[index_name]["mappings"]["_meta"]:
```
Comment (farshidz) [inline src/marqo/tensor_search/backend.py:59]: This doesn't appear to be backwards compatible for existing indexes?

**Review by farshidz — CHANGES_REQUESTED** [review]
In addition to inline comments, please add tests to verify existing indexes will continue to work. I also suggest adding debug level log messges to indicate whether a different search model is being utilised

File: src/marqo/tensor_search/validation.py (hunk 1 above)
Comment (vicilliar) [reply src/marqo/tensor_search/validation.py]: You're right. I had handling in `tensor_search` to convert these `UnknownModelError` to `InvalidArgError`, but it's better to just have them here in the validation function itself.

File: src/marqo/tensor_search/backend.py (hunk 2 above)
Comment (vicilliar) [reply src/marqo/tensor_search/backend.py:59]: fixed, setting `search_model_name` to `None` if it doesn't appear in OS response.

**Comment by vicilliar** [comment]
Added:
1. Debug log message when `search_model` is used for search
2. Tests that `IndexInfo.search_model_name` set to `None` if `search_model` not included in backend mappings response
3. Tests that `determine_model_for_search_vectorisation` returns `model` and `model_properties` if `IndexInfo.search_model_name is None`.
4. Tests that `search` will use `model` for `vectorise` if `search_model` is NOT in backend response.

**Comment by vicilliar** [comment]
Manual tests also added and passed: 
image

**Review by farshidz — APPROVED** [review]

**Pull request merged**